src/tools/perf/perftest.c

```diff
@@ -341,29 +344,31 @@ static ucs_status_t setup_sock_rte_p2p(struct perftest_context *ctx)
     }
 
     if (ctx->server_addr == NULL) {
-        for (t = res; t; t = t->ai_next) {
+        for (t = res; t != NULL; t = t->ai_next) {
             sockfd = socket(t->ai_family, t->ai_socktype, t->ai_protocol);
-            if (sockfd >= 0) {
-                status = ucs_socket_setopt(sockfd, SOL_SOCKET, SO_REUSEADDR,
-                                           &optval, sizeof(optval));
-                if (status != UCS_OK) {
-                    status = UCS_ERR_IO_ERROR;
-                    goto err_close_sockfd;
-                }
+            if (sockfd < 0) {
+                continue;
+            }
 
-                if (!bind(sockfd, t->ai_addr, t->ai_addrlen)) {
-                    break;
-                }
+            status = ucs_socket_setopt(sockfd, SOL_SOCKET, SO_REUSEADDR,
+                                       &optval, sizeof(optval));
+            if (status != UCS_OK) {
+                status = UCS_ERR_IO_ERROR;
+                goto err_close_sockfd;
+            }
 
-                close(sockfd);
-                sockfd = -1;
+            if (bind(sockfd, t->ai_addr, t->ai_addrlen) == 0) {
+                break;
             }
+
+            close(sockfd);
+            sockfd = -2;
         }
 
         if (sockfd < 0) {
-            ucs_error("socket() failed: %m");
+            ucs_error("%s failed: %m", (sockfd == -1) ? "socket()" : "bind()");
```

Comment:
if sockfd is -1, it's also possible getaddrinfo() returned an empty list

Reply:
actually, it is hard to follow the code in case of setting/checking `sockfd`.
maybe it's worth declaring `char err_str[64]` which could be filled by `getaddrinfo() returned empty list`/`socket() failed: %m`/`bind() failed: %m`.
they could be filled by using `snprintf()`, e.g. `snprintf(err_str, 64, "bind() failed: %s", strerror(errno));`

@yosefe wdyt?